### PR TITLE
Add support for responsive image preload.

### PR DESCRIPTION
**Pull request by twifkak**
Compute allowed-alt-sxg directives for URLs in imagesrcset. Behind a
non-default feature until I can confirm that it works correctly in Chromium and
the Google SXG Cache.

Refactor process_link_header a bit to improve readability.

### CODE REVIEW

File: sxg_rs/src/link.rs [hunk 1]
```diff
@@ -14,6 +15,89 @@ pub(crate) async fn process_link_header(
...
+            }
+
+            // If all allowed-alt-sxg directives were found, output the preload and allow
+            // directives.
+            if all_allow_sxg {
+                if let Ok(mut directives) = directives.try_borrow_mut() {
+                    directives.push(link.clone());
+                    directives.extend_from_slice(&allow_directives);
+                }
+            }
```
Comment (twifkak) [inline sxg_rs/src/link.rs:86]: Sorry to combine a new feature and a refactoring. The new production code in this file is in these lines. Let me know if you need me to split it up into 2 commits for easier reading.

File: sxg_rs/src/link.rs [hunk 2]
```diff
@@ -14,6 +15,94 @@ pub(crate) async fn process_link_header(
...
```
Comment (twifkak) [inline sxg_rs/src/link.rs:88]: Sorry to combine a new feature and a refactoring. The new production code in this file is in these lines. Let me know if you need me to split it up into 2 commits for easier reading.

File: sxg_rs/src/link.rs [hunk 3]
```diff
@@ -14,6 +15,94 @@ pub(crate) async fn process_link_header(
...
+            let srcset = get_param(&link.params, "imagesrcset");
+            let images = match &srcset {
+                Some(srcset) => srcset::parse(srcset).unwrap_or_default(),
+                None => vec![],
+            };
+            // Convert image hrefs from srcset to absolute, for comparison with
+            // allowed_alt_sxgs.
+            let images = images
+                .iter()
+                .filter_map(|url| fallback_url.join(url).ok().map(String::from));
```
Comment (antiphoton) [inline sxg_rs/src/link.rs:41]: Does this work when the URL in `srcset` is already an absolute URL?

**Review by antiphoton — APPROVED** [review]

File: sxg_rs/src/link.rs (hunk 3 above)
Comment (twifkak) [reply sxg_rs/src/link.rs:41]: Yes: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8dc6a1bd8b4c0ecbea3a597259f9f0e0

File: sxg_rs/src/http_parser/srcset.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    let mut candidates = vec![];
+    loop {
+        const WHITESPACE: &[char] = &['\t', '\r', '\x0C', '\n', ' '];
+        const WHITESPACE_OR_COMMA: &[char] = &['\t', '\r', '\x0C', '\n', ' ', ','];
+        input = input.trim_start_matches(WHITESPACE_OR_COMMA);
+        if input.is_empty() {
+            return Ok(candidates);
+        }
+        let (mut url, remaining) = input.split_at(input.find(WHITESPACE).unwrap_or(input.len()));
+        println!("remaining: {}", remaining);
```
[the remark this answers is not in the document]
Comment (twifkak) [reply sxg_rs/src/http_parser/srcset.rs]: Done.

File: sxg_rs/src/http_parser/srcset.rs [hunk 5]
```diff
@@ -0,0 +1,96 @@
+#[cfg(feature = "srcset")]
```
[the remark this answers is not in the document]
Comment (twifkak) [reply sxg_rs/src/http_parser/srcset.rs:15]: Done.

File: .github/workflows/unit-tests.yml [hunk 6]
```diff
@@ -47,6 +47,7 @@ jobs:
       - uses: actions-rs/cargo@v1
         with:
           command: test
+          args: --features srcset
```
[the remark this answers is not in the document]
Comment (twifkak) [reply .github/workflows/unit-tests.yml]: Done.

**Comment by twifkak** [comment]
Thanks for the review! I'll merge as soon as the checks pass after these changes.

**Pull request merged**